### PR TITLE
Homework #4 / repository, providers, logging

### PR DESCRIPTION
**Pull request by delveper**
# Summary

This pull request introduces significant improvements to our project, primarily focusing on the integration of exchange rate provider and enhancements to the logger package.

## Repository

The Repository pattern was implemented on early stages of the project to provide a cleaner, more encapsulated, and reusable way of handling data operations. This pattern allows us to abstract the details of data access and manipulation, making the code more maintainable and testable. The [`subs/repo.go`](https://github.com/GenesisEducationKyiv/main-project-delveper/blob/hw4-repos-provs-logs/internal/subs/repo.go) the implementation of the Repository pattern.

## Providers

The `curxrt` package is a crucial part of our project. It contains the implementation of different currency exchange rate providers such as alphavantage, coinapi, coinyep, and others. Each provider has its own dedicated file, and they all implement the `Provider` abstraction defined in [`curxrt/curxrt.go`](https://github.com/GenesisEducationKyiv/main-project-delveper/blob/hw4-repos-provs-logs/internal/rate/curxrt/curxrt.go). This design allows for easy addition or removal of providers. Using generics is debatable, but in this particular case, the code remains DRY. However, this design has a drawback: it lacks readability and pollutes the interface map.

## Logger

The logger package has been enhanced to provide more detailed logging information using console and file logging. The `logProviderEvent` function dealing with **event** and logs the events generated by the provider. 

## Event 
Despite `event` package out of scope HW4 it has been integrated into the project to improve the overall functionality. This package designed to provide handling system events that allows different parts of the application to communicate with each other without being tightly coupled. The [`event/event.go` ](https://github.com/GenesisEducationKyiv/main-project-delveper/blob/hw4-repos-provs-logs/sys/event/event.go) file contains the implementation of the event system. **It's important to note that the current implementation of the event package is a **prototype** and is primarily used to log provider responses.**

### CODE REVIEW

File: internal/rate/curxrt/ninjas.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+func (p Ninjas) ProcessResponse(resp *http.Response) (float64, error) {
+	var data struct {
+		CurrencyPair string  `json:"currency_pair"`
+		ExchangeRate float64 `json:"exchange_rate"`
+	}
+
+	if err := web.ProcessResponse(resp, &data); err != nil {
+		return 0, err
```
Comment (lumenalux) [inline internal/rate/curxrt/ninjas.go:28]: Same

File: internal/rate/curxrt/xratehost.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			URL string `json:"url"`
+		} `json:"motd"`
+		Success bool               `json:"success"`
+		Base    string             `json:"base"`
+		Date    string             `json:"date"`
+		Rates   map[string]float64 `json:"rates"`
+	}
+
+	if err := web.ProcessResponse(resp, &data); err != nil {
+		return 0, err
```
Comment (lumenalux) [inline internal/rate/curxrt/xratehost.go:34]: Same

File: sys/web/respond.go [hunk 3]
```diff
@@ -30,33 +30,41 @@ func Respond(ctx context.Context, rw http.ResponseWriter, data any, code int) er
 	return nil
 }
```
Comment (lumenalux) [inline sys/web/respond.go:31]:
```go
// Respond responds with converted data to the client.
func Respond(ctx context.Context, rw http.ResponseWriter, data any, code int) error {
	if code == http.StatusNoContent {
		rw.WriteHeader(code)
		return nil
	}

	var buf bytes.Buffer
	if err := EncodeBody(&buf, data); err != nil {
		return fmt.Errorf("encoding to buffer: %w", err)
	}

	rw.WriteHeader(code)
	if _, err := buf.WriteTo(rw); err != nil {
		return fmt.Errorf("writing response: %w", err)
	}
	return nil
}
```
Is context used here? Is it necessary for this function to work?

File: sys/web/respond.go [hunk 4]
```diff
@@ -30,33 +30,41 @@ func Respond(ctx context.Context, rw http.ResponseWriter, data any, code int) er
...
-	if err := json.NewDecoder(body).Decode(data); err != nil {
-		return fmt.Errorf("decoding body: %w", err)
+// EncodeBody writes data to a writer after converting it to JSON.
+func EncodeBody(rw io.Writer, data any) error {
+	if err := json.NewEncoder(rw).Encode(data); err != nil {
+		return fmt.Errorf("encoding body: %w", err)
 	}
 
-	return err
+	return nil
```
Comment (lumenalux) [inline sys/web/respond.go:56]: Explicitly returning `nil` as an error instead of `err` - :+1:

**Review by lumenalux — COMMENTED** [review]
I noticed that code tends to utilize generics. Generics can provide code with flexibility and reduce boilerplate. However, it's important to consider whether they are necessary for the project. While generics offer pros, they also come with cons. If a significant portion of the code incorporates generics from the start, it can potentially lead to performance issues as the project grows in complexity

https://www.infoq.com/news/2022/04/go-generics-performance/

https://planetscale.com/blog/generics-can-make-your-go-code-slower

File: cmd/main.go [hunk 5]
```diff
@@ -11,13 +11,13 @@ import (
...
 	"github.com/GenesisEducationKyiv/main-project-delveper/internal/rate"
-	"github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription"
+	"github.com/GenesisEducationKyiv/main-project-delveper/internal/subs"
 	"github.com/GenesisEducationKyiv/main-project-delveper/sys/env"
 	"github.com/GenesisEducationKyiv/main-project-delveper/sys/logger"
 )
 
 func main() {
-	log := logger.New(logger.LevelDebug)
+	log := logger.New(logger.LevelDebug, "./log/sys.log")
```
[the remark this answers is not in the document]
Comment (kachan1208) [reply cmd/main.go:20]: prefer to use stdout/stderr.

File: internal/subs/event.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+const (
+	EventSource          = "subs"
+	EventKindRateRequest = "rate_request"
+)
+
+var ErrInvalidEvent = errors.New("invalid event")
+
+type RequestEvent interface {
```
Comment (kachan1208) [inline internal/subs/event.go:18]: prefer embed at least like this: https://go.dev/doc/effective_go#embedding

File: internal/subs/subs.go [hunk 7]
```diff
@@ -96,23 +104,25 @@ func (svc *Service) Subscribe(ctx context.Context, sub Subscriber) error {
...
-
-	rate, err := svc.rate.Get(ctx, pair)
+	subscribers, err := svc.repo.List(ctx)
 	if err != nil {
 		return err
 	}
 
-	subscribers, err := svc.repo.List(ctx)
+	// TODO: Improve the logic of retrieving exchange rate for distinct topics.
+	topic := subscribers[0].Topic
```
Comment (kachan1208) [inline internal/subs/subs.go:113]: what if there is no subscribers in the list?

File: internal/subs/subs.go [hunk 8]
```diff
@@ -96,23 +104,25 @@ func (svc *Service) Subscribe(ctx context.Context, sub Subscriber) error {
...
+
+	val, err := svc.RequestExchangeRate(ctx, topic)
```
Comment (kachan1208) [inline internal/subs/subs.go:115]:
```suggestion
	exchangeRate, err := svc.RequestExchangeRate(ctx, topic)
```

File: internal/rate/event.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	return rd.quoteCurrency
+}
+
+// ExchangeRate returns the exchange rate of the ResponseEventData.
+func (rd *ResponseEventData) ExchangeRate() float64 {
+	return rd.exchangeRate
+}
+
+// RespondExchangeRate handles an event and fetches the exchange rate for a requested currency pair.
+func (svc *Service) RespondExchangeRate(ctx context.Context, e event.Event) error {
```
Comment (kachan1208) [inline internal/rate/event.go:65]: isn't it better to use event.RequestEvent type there?

File: internal/subs/event.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	req := event.New(EventSource, EventKindRateRequest, toRequestEventData(topic))
+	ch := make(chan event.Event)
+	req.Response = ch
+
+	if err := svc.bus.Dispatch(ctx, req); err != nil {
+		return 0, err
+	}
+
+	select {
+	case <-ctx.Done():
```
Comment (kachan1208) [inline internal/subs/event.go:68]: unit tests pls.

File: internal/subs/handler.go [hunk 11]
```diff
@@ -61,10 +61,14 @@ func (h *Handler) Subscribe(ctx context.Context, rw http.ResponseWriter, req *ht
...
 
-	sub := toSubscriber(email, "") // TODO: add topic to subscriber.
+	// TODO: Fetch a topic from query params or JSON body.
+	sub := toSubscriber(email, NewTopic(currencyBTC, currencyUAH))
 	if err := h.SubscriptionService.Subscribe(ctx, sub); err != nil {
-		if errors.Is(err, ErrEmailAlreadyExists) {
+		switch {
+		case errors.Is(err, ErrEmailAlreadyExists):
 			return web.NewRequestError(err, http.StatusConflict)
+		case errors.Is(err, context.DeadlineExceeded):
```
Comment (kachan1208) [inline internal/subs/handler.go:70]: unit tests pls

File: sys/web/middlewares_test.go [hunk 12]
```diff
@@ -59,7 +59,7 @@ func TestChainMiddlewares(t *testing.T) {
 }
 
 func TestMiddlewares(t *testing.T) {
-	log := logger.New(logger.LevelDebug)
+	log := logger.New(logger.LevelDebug, "../../log/test.log")
```
Comment (kachan1208) [inline sys/web/middlewares_test.go:62]: Why not stdout/stderr?

File: sys/web/request.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
+package web
```
Comment (kachan1208) [inline sys/web/request.go:1]: like for option pattern.
Great usage.

File: sys/web/respond.go [hunk 14]
```diff
@@ -30,33 +30,41 @@ func Respond(ctx context.Context, rw http.ResponseWriter, data any, code int) er
...
+	dec := json.NewDecoder(body)
+	dec.DisallowUnknownFields()
+
+	if err := dec.Decode(data); err != nil {
+		return fmt.Errorf("decoding body: %w", err)
 	}
 
-	if val, ok := data.(interface{ OK() error }); !ok {
-		return fmt.Errorf("validation: %w", val.OK())
+	if val, ok := data.(interface{ Validate() error }); ok {
```
Comment (kachan1208) [inline sys/web/respond.go:43]: don't you violate Single responsibility pattern there?

**Review by kachan1208 — APPROVED** [review]

**Comment by kachan1208** [comment]
In general a lot of integrations, but lack of tests.

**Comment by kachan1208** [comment]
Consider adding tests for exchange rates using httptest.

File: cmd/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (delveper) [reply cmd/main.go:20]: I log to both the console (stderr) and a JSON file.

File: internal/subs/subs.go (hunk 7 above)
Comment (delveper) [reply internal/subs/subs.go:113]: 109 will produce `os.ErrNotExist` from file store

File: sys/web/middlewares_test.go (hunk 12 above)
Comment (delveper) [reply sys/web/middlewares_test.go:62]: I log to stderr and file

File: sys/web/respond.go (hunk 14 above)
Comment (delveper) [reply sys/web/respond.go:43]: I am guilty. It is debatable where validation should happen. I think when we  able to do it right away it is justified. Do you prefer to do it on the business logic side?

File: internal/subs/event.go (hunk 6 above)
Comment (delveper) [reply internal/subs/event.go:18]: Sorry I do not get it. Should I change interface name? I am doing literally the same: 
```go
type ReadWriter interface {
    Reader
    Writer
}
```

Comment (delveper) [reply internal/subs/event.go:18]: Or problem in this double wrapping:
```go
type ResponseEvent interface {
	ExchangeRate
}
```

File: sys/web/respond.go (hunk 14 above)
Comment (kachan1208) [reply sys/web/respond.go:43]: Yep. Decode data should do just the one thing at a time.

**Pull request merged**